cmd/crank/beta/trace/internal/resource/client.go

```diff
@@ -0,0 +1,210 @@
+	if err != nil {
+		return nil, errors.Wrap(err, errCouldntGetRootResource)
+	}
+
+	// breadth-first search of children
+	queue := []*Resource{root}
+
+	for len(queue) > 0 {
+		res := queue[0]   // Take the first element
+		queue = queue[1:] // Dequeue the first element
+
+		refs := getResourceChildrenRefs(res, kc.getConnectionSecrets)
+		if err != nil {
+			return nil, errors.Wrap(err, errCouldntGetRootResource)
+		}
+		for i := range refs {
+			child, err := kc.getResource(ctx, &refs[i])
+			if err != nil {
+				return nil, errors.Wrap(err, errCouldntGetChildResource)
+			}
+			res.Children = append(res.Children, child)
+			queue = append(queue, child) // Enqueue child
+		}
+	}
+
+	return root, nil
+}
+
+// getResource returns the requested Resource.
+func (kc *Client) getResource(ctx context.Context, ref *v1.ObjectReference) (*Resource, error) {
+	result := unstructured.Unstructured{}
+	result.SetGroupVersionKind(ref.GroupVersionKind())
+	err := kc.client.Get(ctx, xpmeta.NamespacedNameOf(ref), &result)
+	if kerrors.IsNotFound(err) {
+		result.SetName(ref.Name)
+		result.SetNamespace(ref.Namespace)
+		return &Resource{Unstructured: result, Error: err}, nil
+	}
+	if err != nil {
+		return nil, errors.Wrap(err, errCouldntGetResource)
+	}
+
+	return &Resource{Unstructured: result}, nil
+}
+
+// getResourceChildrenRefs returns the references to the children for the given
+// Resource, assuming it's a Crossplane resource, XR or XRC.
+func getResourceChildrenRefs(r *Resource, getConnectionSecrets bool) []v1.ObjectReference {
+	obj := r.Unstructured
+	// collect owner references
+	var refs []v1.ObjectReference
+
+	xrc := claim.Unstructured{Unstructured: obj}
+	if ref := xrc.GetResourceReference(); ref != nil {
+		refs = append(refs, v1.ObjectReference{
+			APIVersion: ref.APIVersion,
+			Kind:       ref.Kind,
+			Name:       ref.Name,
+			Namespace:  ref.Namespace,
+		})
+	}
+
+	xr := composite.Unstructured{Unstructured: obj}
+	refs = append(refs, xr.GetResourceReferences()...)
+
+	if !getConnectionSecrets {
+		return refs
+	}
+
+	// handle write connection secrets for both XR and XRC
+	xrSecretRef := xr.GetWriteConnectionSecretToReference()
+	if xrSecretRef != nil {
+		ref := v1.ObjectReference{
+			APIVersion: "v1",
+			Kind:       "Secret",
+			Name:       xrSecretRef.Name,
+			Namespace:  xrSecretRef.Namespace,
+		}
+		if ref.Namespace == "" {
+			// this was actually an XRC secret ref, which are local, so we need
+			// to use the XRC's namespace
+			ref.Namespace = r.Unstructured.GetNamespace()
+		}
+		refs = append(refs, ref)
+	}
+
+	return refs
+}
+
+// MappingFor returns the RESTMapping for the given resource or kind argument.
+// Copied over from cli-runtime pkg/resource Builder,
+// https://github.com/kubernetes/cli-runtime/blob/9a91d944dd43186c52e0162e12b151b0e460354a/pkg/resource/builder.go#L768
+func (kc *Client) MappingFor(resourceOrKindArg string) (*meta.RESTMapping, error) {
+	// TODO(phisco): actually use the Builder.
```

Comment:
it's more a wish, there is no real need. It would be awesome to be able to leverage the upstream code, but it shouldn't change the end result, so I wouldn't actively track it.